composer/callbacks/run_directory_uploader.py

```diff
@@ -0,0 +1,319 @@
+        container=container,
+        object_name=f"{object_name_prefix}.validate_credentials_success",
+    )
+
+
+def _upload_worker(
+    file_queue: Union[queue.Queue[str], multiprocessing.JoinableQueue[str]],
+    is_finished: Union[multiprocessing._EventType, threading.Event],
+    upload_staging_dir: str,
+    provider_name: str,
+    container_name: str,
+    object_name_prefix: str,
+    init_kwargs: Dict[str, Any],
+):
+    """A long-running function to handle uploading files.
+
+    Args:
+        file_queue (queue.Queue or multiprocessing.JoinableQueue): The worker will poll
+            this queue for files to upload.
+        is_finished (threading.Event or multiprocessing.Event): An event that will be
+            set when training is finished and no new files will be added to the queue.
+            The worker will continue to upload existing files that are in the queue.
+            When the queue is empty, the worker will exit.
+        upload_staging_dir (str): The upload staging directory.
+        provider_name (str): The cloud provider name.
+        container_name (str): The container name (e.g. s3 bucket) for the provider
+            where files will be uploaded.
+        object_name_prefix (str): Prefix to prepend to the object names
+             before they are uploaded to the blob store.
+        init_kwargs (Dict[str, Any]): Arguments to pass in to the
+            :class:`~libcloud.storage.providers.Provider` constructor.
+    """
+    from libcloud.common.types import LibcloudError
+    from libcloud.storage.providers import get_driver
+    provider_cls = get_driver(provider_name)
+    provider = provider_cls(**init_kwargs)
+    container = provider.get_container(container_name)
+    while True:
+        try:
+            file_path_to_upload = file_queue.get(block=True, timeout=0.5)
+        except queue.Empty:
+            if is_finished.is_set():
+                break
+            else:
+                continue
+        obj_name = ",".join(os.path.relpath(file_path_to_upload, upload_staging_dir).split(
+            os.path.sep)[1:])  # the first folder is the upload timestamp. Chop that off.
+        log.info("Uploading file %s to %s://%s/%s%s", file_path_to_upload, provider_name, container_name,
+                 object_name_prefix, obj_name)
+        retry_counter = 0
+        while True:
+            try:
+                provider.upload_object(
+                    file_path=file_path_to_upload,
+                    container=container,
+                    object_name=object_name_prefix + obj_name,
+                )
+            except LibcloudError as e:
+                # The S3 driver does not encode the error code in an easy-to-parse manner
+                # So doing something fairly basic to retry on transient error codes
+                if any(x in str(e) for x in ("408", "409", "425", "429", "500", "503", '504')):
```

Comment:
Should 500 be included in transient errors? It probably shouldn't

Reply:
500s are the catchall...would rather retry than not, in case if it is transient like a race